src/main/java/io/r2dbc/postgresql/PostgresqlConnectionFactoryProvider.java

```diff
@@ -48,6 +54,16 @@
      */
     public static final Option<String> SCHEMA = Option.valueOf("schema");
 
+    /**
+     * Connection options which are applied once after the connection has been created.
+     * Format is '{key}={value},...,{key}={value}'
+     */
+    public static final Option<String> OPTIONS = Option.valueOf("options");
```

Comment:
Introducing `options` contradicts somewhat the purpose of the query-string/`ConnectionFactoryOptions ` itself. The idea of options is to keep config parameters on the top-level. I'd suggest to include a collection of known option names and then to iterate over these instead of extracting an options object.

Reply:
That's true, but there are already tons of such options out there and each new release of PostgreSQL might introduce new options. Introducing constants for known option names introduces two problems IMO:

- it would require a new release of r2dbc-postgresql for new options introduced in PostgreSQL which adds a dependency one might not want. It requires you to update r2dbc-postgresql if you upgrade your DB server and want to use new options.
- the availability of these options as constants might let users think that they are generally available. Activating a (let's say) PG11 option in the ConnectionFactoryOptions but using a PG9 server will fail badly, possibly irritating the user. You would explicitly have to make it absolutely clear in the documentation which releases of PG are compatible. Combined with the hundreds of possible options already available… I don't know if this is a good idea.

Reply:
PGJDBC has also an options object. Would it make sense to align R2DBC Postgres' `options` to what JDBC does? JDBC transports options in the startup packet which removes the need of running subsequent `SET` statements.

Reply:
I was able to modify the code to set the start options in the `StartupMessage` using the PGJDBC code as a template as proposed. Then I checked the official PostgreSQL documentation at https://www.postgresql.org/docs/current/protocol-message-formats.html which reads:

> options Command-line arguments for the backend. (This is deprecated in favor of setting individual run-time parameters.) 

So yes, I could move the connection initialisation parameters to the `StartupMessage`, but this is regarded as a deprecated way. What do you suggest? Go ahead moving it to `StartupMessage` or keep the implementation as-is?

Reply:
Thanks a lot. For database-specifics, I’m fan of doing what JDBC does. Maybe @davecramer can hint us in the right direction.

Reply:
I don't see anywhere where the jdbc driver uses options. It sends the parameters as parameters. https://github.com/pgjdbc/pgjdbc/blob/fcbbc3e6408cc1bcf459b740c683f3db40a5050c/pgjdbc/src/main/java/org/postgresql/core/v3/ConnectionFactoryImpl.java#L450

Reply:
And `options` is one of these parameters. It is pulled from the properties in lines 343-346 and merged into the parameters. There is a test case for this feature as well (https://github.com/pgjdbc/pgjdbc/blob/fcbbc3e6408cc1bcf459b740c683f3db40a5050c/pgjdbc/src/test/java/org/postgresql/test/core/OptionsPropertyTest.java). 

But as mentioned above, the `options` field is described as deprecated in the official PG docs.

Reply:
Fair enough, although deprecated does not mean it will be going away anytime soon. That said not using options is probably better

Reply:
Not using options means keeping the current implementation, i.e., sending a series of `SET` statements after creating the connection?

Reply:
Why send set for each option? The startup message can just contain all the parameters in one message.

Reply:
I think I do not fully get what you propose. You suggest setting the parameters in the startup message, but not to use options. But this (deprecated) options field in the startup message would be where to set them according to the protocol. This is also how it is implemented in pgjdbc.

Reply:
It is deprecated in favour of just sending parameters. You send name, value. JDBC encodes these into UTF-8 and each name, values is sent as a pair with a trailing 0. Options is not used. There is no need for multiple SET commands. See https://github.com/pgjdbc/pgjdbc/blob/fcbbc3e6408cc1bcf459b740c683f3db40a5050c/pgjdbc/src/main/java/org/postgresql/core/v3/ConnectionFactoryImpl.java#L476-L486

Reply:
So basically sending `options` as `-c search_path=foo` translates to a parameter key-value tuple with `search_path` as name and `foo` as value?

Reply:
Essentially, yes. And yes it is sent as a tuple.

Reply:
Ok, I think I get it now. :) I'll modify the code accordingly and push it for further review.